pymc/sampling/forward.py

```diff
@@ -497,22 +499,148 @@ def sample_posterior_predictive(
 
     Examples
     --------
-    Thin a sampled inferencedata by keeping 1 out of every 5 draws
-    before passing it to sample_posterior_predictive
+    The behavior of ``var_names``
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    The function returns posterior predictive samples of any variable included in ``var_names``,
+    conditioned on the values of other random variables found in the trace.
+    To ensure the samples are internally consistent, any random variable that depends
+    on another random variable that is being sampled is itself sampled.
+    The final list of variables being sampled is shown in the log output.
+
+    Note that if a random variable has no dependency on other random variables,
+    posterior predictive sampling is equivalent to prior sampling.
+    Likewise, if all random variables are being sampled, the behavior is equivalent to
+    that of :func:`~pymc.sample_prior_predictive`.
+
+    .. warning:: A random variable included in ``var_names`` will never be copied from the posterior group. It will always be sampled as described above. If you want, you can copy manually via ``idata.posterior_predictive["var_name"] = idata.posterior["var_name"]``.
+
+
+    The following code block explores how the behavior changes with different var_names:
+
+    .. code:: python
+
+        from logging import getLogger
+        import pymc as pm
+
+        # Some environments like google colab suppress
+        # the default logging output of PyMC
+        getLogger("pymc").setLevel("INFO")
+
+        kwargs = {"progressbar": False, "random_seed": 0}
+
+        with pm.Model() as model:
+          x = pm.Normal("x")
+          y = pm.Normal("y")
+          z = pm.Normal("z", x + y**2)
+          det = pm.Deterministic("det", pm.math.exp(z))
+          obs = pm.Normal("obs", det, 1, observed=[20])
+
+          idata = pm.sample(tune=10, draws=10, chains=2, **kwargs)
+
+    Default behavior. Posterior predictive draws of ``obs``,
+    conditioned on the posterior draws of ``z`` found in the trace.
+
+    .. code:: python
+
+          pm.sample_posterior_predictive(idata, var_names=["obs"], **kwargs)
+          # Sampling: [obs]
+
+    Re-compute the deterministic variable ``det``,
+    conditioned on the posterior draws of ``z`` found in the trace.
+
+    .. code :: python
+
+          pm.sample_posterior_predictive(idata, var_names=["det"], **kwargs)
+          # Sampling: []
+
+    Posterior predictive draws of ``z`` and ``det``,
```

Comment:
I would not call those posterior predictive draws. The most similar (and "sensible"?) case that comes to mind is predicting **parameters** for unseen groups in hierarchical models, which to me would be posterior samples. i.e. the alphas in https://mc-stan.org/docs/stan-users-guide/mixed-replication.html which don't have a name there either.

Happy to try and find/come up with a name for these kind of quantities but would avoid calling them posterior predictive (even if generated by `sample_posterior_predictive`) as this is one of the key points of all the discussion imo, `sample_posterior_predictive` doesn't necessarly generate posterior predictive samples

Reply:
Afaict there is no name for this. I still don't buy the distinction, it's a prediction and it's based on the posterior. Whether you could ever observe this variable or not sounds a bit too philosophical.

Reply:
> sampling needs the logp which needs to evaluate the different contributions, some as priors others as likelihood

The distinction doesn't matter for mcmc :)

I am happy to use another name if one can be suggested, but I think this really touches on the issue of behavior vs intention. I am more interested in explaining the behavior, because the intention will always depend on the user.

It's like if we worried (too much) about not confusing users in the docstrings of `pm.sample` when we show models without observed variables. Are they posterior or prior samples? But then we can philosophize that a prior is a posterior when data is empty.

Reply:
> docstrings of pm.sample when we show models without observed variables. Are they posterior or prior (+- predictive adjective) samples?

I wasn't aware of this, but it can be fine if the focus is on behaviour and the result is not called "posterior".

We can rewrite the text here to avoid using "posterior predictive" maybe even give that quantity a name, and focusing on behaviour is helpful (and in my opinion also prevents us from using posterior predictive here as posterior predictive can't be disentangled with intent). We could use something like:

```
Generate draws from the distributions of ``z`` and ``det``,
conditioned on the draws of ``x``` and ``y``` found in `trace`.
```

I have also removed "posterior" and "the trace" to reference only `` `trace` `` aka the first argument of the function